src/constructs/aws/Upload.ts

```diff
@@ -0,0 +1,179 @@
+import { BlockPublicAccess, Bucket, BucketEncryption, HttpMethods } from "@aws-cdk/aws-s3";
+import type { Construct as CdkConstruct } from "@aws-cdk/core";
+import { CfnOutput, Duration, Fn, Stack } from "@aws-cdk/core";
+import { Code, Function as LambdaFunction, Runtime } from "@aws-cdk/aws-lambda";
+import type { FromSchema } from "json-schema-to-ts";
+import type { AwsProvider } from "@lift/providers";
+import { AwsConstruct } from "@lift/constructs/abstracts";
+import type { IHttpApi } from "@aws-cdk/aws-apigatewayv2";
+import { HttpApi, HttpMethod, HttpRoute, HttpRouteKey } from "@aws-cdk/aws-apigatewayv2";
+import type { Resource } from "@aws-cdk/aws-apigateway";
+import { LambdaIntegration, RestApi } from "@aws-cdk/aws-apigateway";
+import { LambdaProxyIntegration } from "@aws-cdk/aws-apigatewayv2-integrations";
+import { Role } from "@aws-cdk/aws-iam";
+import { PolicyStatement } from "../../CloudFormation";
+
+const UPLOAD_DEFINITION = {
+    type: "object",
+    properties: {
+        type: { const: "upload" },
+        apiGateway: { enum: ["http", "rest"] },
+        encryption: {
+            anyOf: [{ const: "s3" }, { const: "kms" }],
+        },
+    },
+    additionalProperties: false,
+} as const;
+const UPLOAD_DEFAULTS: Required<FromSchema<typeof UPLOAD_DEFINITION>> = {
+    type: "upload",
+    encryption: "s3",
+    apiGateway: "http",
+};
+
+type Configuration = FromSchema<typeof UPLOAD_DEFINITION>;
+
+export class Upload extends AwsConstruct {
+    public static type = "upload";
+    public static schema = UPLOAD_DEFINITION;
+
+    private readonly bucket: Bucket;
+    private readonly bucketNameOutput: CfnOutput;
+    private function: LambdaFunction;
+    private httpApi: IHttpApi | undefined;
+    private route: HttpRoute | undefined;
+    private restApi: RestApi | undefined;
+
+    constructor(scope: CdkConstruct, id: string, configuration: Configuration, private provider: AwsProvider) {
+        super(scope, id);
+
+        const resolvedConfiguration = Object.assign({}, UPLOAD_DEFAULTS, configuration);
+
+        const encryptionOptions = {
+            s3: BucketEncryption.S3_MANAGED,
+            kms: BucketEncryption.KMS_MANAGED,
+        };
+
+        this.bucket = new Bucket(this, "Bucket", {
+            encryption: encryptionOptions[resolvedConfiguration.encryption],
+            blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
+            enforceSSL: true,
+            cors: [
+                {
+                    allowedMethods: [HttpMethods.PUT],
+                    allowedOrigins: ["*"],
```

Comment:
for this first version I chose to set wide CORS permissions. We may add an option to configure it later.